boa/src/syntax/parser/cursor/mod.rs

```diff
@@ -0,0 +1,186 @@
+//! Cursor implementation for the parser.
+mod buffered_lexer;
+
+use super::ParseError;
+use crate::syntax::{
+    ast::Punctuator,
+    lexer::{InputElement, Lexer, Position, Token, TokenKind},
+};
+use buffered_lexer::BufferedLexer;
+use std::io::Read;
+
+/// Token cursor.
+///
+/// This internal structure gives basic testable operations to the parser.
+#[derive(Debug)]
+pub(super) struct Cursor<R> {
+    buffered_lexer: BufferedLexer<R>,
+}
+
+impl<R> Cursor<R>
+where
+    R: Read,
+{
+    /// Creates a new cursor.
+    #[inline(always)]
+    pub(super) fn new(reader: R) -> Self {
+        Self {
+            buffered_lexer: Lexer::new(reader).into(),
+        }
+    }
+
+    #[inline(always)]
+    pub(super) fn set_goal(&mut self, elm: InputElement) {
+        self.buffered_lexer.set_goal(elm)
+    }
+
+    #[inline(always)]
+    pub(super) fn lex_regex(&mut self, start: Position) -> Result<Token, ParseError> {
+        self.buffered_lexer.lex_regex(start)
+    }
+
+    #[inline(always)]
+    pub(super) fn next(
+        &mut self,
+        skip_line_terminators: bool,
```

Comment:
Currently, in the majority of cases, we do not skip line terminators - but as you say in most cases we should be.

The problem is described above but basically there is an issue that if we peek(true) then we delete a line terminator which might be needed later. I think (but I haven't checked yet though) that your suggestion above (collapse line terminators and then look at index+1) will solve this but I need to check as it might lead to the possibility where a line terminator is left 'stuck' in the buffer longer than it should which later causes a problem.